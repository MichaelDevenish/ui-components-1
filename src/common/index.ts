export { Props } from './types'
export { getColor } from './legacy'
