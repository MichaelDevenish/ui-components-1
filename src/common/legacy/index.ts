export { getColor } from './colors/getColor'
