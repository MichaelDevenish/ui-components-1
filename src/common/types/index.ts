/**
 * For Examples, see:
 * https://github.com/Microsoft/TypeScript/blob/master/src/compiler/types.ts
 * https://github.com/alm-tools/alm/blob/master/src/common/types.ts
 */
import { Size as _s, AllSizes } from './size'

namespace Props {
  export import Size = _s
}

export {
  Props,
  AllSizes
}
