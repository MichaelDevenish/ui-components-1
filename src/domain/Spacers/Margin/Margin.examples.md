Margins is a prop that can be provided to some components that will define the margin they
have around them. It can be used as a cleaner alternative to layoutSpacer for simple component
arrangements. 

#### Simple margins

Buttons are one of the components that support margins. As seen below when you provide the
margins prop with any combination of top, bottom, left, or right it will apply the spacing
to the component.

```jsx
import { Variables } from '@Common';
import { Button } from '@Domain/Buttons';

<>
  <Button
    margins={{
      top: Variables.Spacing.sSmall,
      bottom: Variables.Spacing.sLarge,
      left: Variables.Spacing.sSmall,
      right: Variables.Spacing.sLarge,
    }}
  >
    Spaced Button
  </Button>
  <Button
    margins={{
      top: Variables.Spacing.sSmall,
      right: Variables.Spacing.sSmall
    }}
  >
    Spaced Button
  </Button>
</>
```


#### Breakpoint margins

You can also provide breakpoints for margins so that it can have a different spacing on
mobile vs desktop.

```jsx
import { Variables } from '@Common';
import { Button } from '@Domain/Buttons';

<Button
  margins={{
    top: {
      min: Variables.Spacing.sLarge,
      tablet: 'none',
      desktop: Variables.Spacing.sSmall
    },
    bottom: Variables.Spacing.sLarge,
    left: Variables.Spacing.sSmall,
    right: Variables.Spacing.sLarge,
  }}
>
  Spaced Button
</Button>
```

#### Margin Component

There is also a standard margin component which can be used when trying 
to use margins on components that do not support margin props.

```jsx
import { Variables } from '@Common';

<div style={{ backgroundColor: Variables.Color.n400, overflow: 'auto', width: 'fit-content', height: 'fit-content' }}>
    <Margin
      margins={{
        top: Variables.Spacing.sXLarge,
        right: Variables.Spacing.sXLarge,
        left: Variables.Spacing.sXLarge,
        bottom: Variables.Spacing.sXLarge
      }}
    >
      <div style={{ height: 200, width: 200, backgroundColor: Variables.Color.n600 }}/>
    </Margin>
</div>
```
