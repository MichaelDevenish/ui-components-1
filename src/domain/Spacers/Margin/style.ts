import styled from 'styled-components'

import { Props } from '../../../common'
import { styleForMargins } from '../services/margins'

const StyledMargin = styled.div<{margins?: Props.IMargins}>`
  ${({margins}) => styleForMargins(margins)}
`

export {
  StyledMargin
}
