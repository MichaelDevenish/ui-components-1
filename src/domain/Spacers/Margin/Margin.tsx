import { isEqual } from 'lodash'
import React from 'react'

import { Props } from '../../../common'
import { StyledMargin } from './style'

const Margin: React.FC<{
  margins?: Props.IMargins
  children?: React.ReactNode
}> = ({margins, children}) =>  (
  <StyledMargin
    margins={margins}
  >
    {children}
  </StyledMargin>
)

const MemoMargin = React.memo(Margin, (prevProps, nextProps) => isEqual(prevProps.margins, nextProps.margins))
MemoMargin.displayName = 'Margin'

export {
  MemoMargin as Margin
}
