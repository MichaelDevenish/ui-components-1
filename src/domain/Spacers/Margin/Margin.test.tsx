import { shallow } from 'enzyme'
import React from 'react'

import { Variables } from '../../../common'
import { Margin } from './Margin'

describe('<Margin />', () => {
  describe('Standard Margin', () => {
    const wrapper = shallow(
      <Margin
        margins={{
          top: {
            min: Variables.Spacing.sLarge,
            tablet: 'none',
            desktop: Variables.Spacing.sSmall
          },
          bottom: Variables.Spacing.sLarge,
          left: Variables.Spacing.sSmall,
          right: Variables.Spacing.sLarge
        }}
      >
        <div>Margin</div>
      </Margin>
    )

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    })
  })
})
