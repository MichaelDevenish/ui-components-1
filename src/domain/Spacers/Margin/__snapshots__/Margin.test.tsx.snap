// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Margin /> Standard Margin should match the snapshot 1`] = `
<styled.div
  margins={
    Object {
      "bottom": 24,
      "left": 12,
      "right": 24,
      "top": Object {
        "desktop": 12,
        "min": 24,
        "tablet": "none",
      },
    }
  }
>
  <div>
    Margin
  </div>
</styled.div>
`;
